File: packages/ui-shared/src/types/media.ts
// Media shapes shared by the Control Plane admin UI and the Agent
// Dashboard. Both render the same normalized payload, so both must agree
// on what a media element is — two parallel copies of this shape is the
// drift that made audio and PDFs render as zero-byte images.

/** Which kind of thing the bytes are. Never inferred from the block type. */
export type MediaModality = 'image' | 'audio' | 'video' | 'file';

/**
 * Custody: whether the bytes exist, and if not, why. Only `captured` has
 * reachable bytes; every other state must render an explanation rather
 * than a control that would fail when clicked.
 */
export type MediaSource =
  /** Held in the governed body, addressable via `locator`. */
  | 'captured'
  /** Never held by design — a remote URL the gateway never fetched. */
  | 'external'
  /** Never held by design — the provider holds it under its own id. */
  | 'provider-ref'
  /** Bytes were not retained; a sha256/size/mime stamp survives. */
  | 'fingerprint'
  /**
   * Captured once, then removed by retention. Produced by the Control
   * Plane when it serves a stored sidecar after establishing that no body
   * is recoverable — the digest survives so the file is still identifiable,
   * the locator does not, so no control is offered for bytes that are gone.
   */
  | 'aged-out'
  /** Never captured at all. */
  | 'absent';

/**
 * One media element of a captured exchange. It never carries content
 * bytes in any encoding: `locator` addresses them inside the captured
 * body, and the app's byte resolver dereferences it.
 */
export interface MediaRef {
  modality: MediaModality;
  /**
   * Wire-declared type. A display hint ONLY — a caller controls the wire,
   * so this must never be used to construct a Blob type. The bytes' real
   * type comes from the resolver.
   */
  mime?: string;
  /**
   * Decoded size in bytes. Always present — a zero-byte element and an
   * element of unknown size are different facts and must stay
   * distinguishable.
   */
  sizeBytes: number;
  source: MediaSource;
  /** Set only when source is `captured`. */
  locator?: string;
  /** Set only when source is `external`. Inert — never dereferenced. */
  url?: string;
  /** Set only when source is `provider-ref`. */
  providerRef?: string;
  /** A real hex digest, or absent. */
  sha256?: string;
  /** Bytes are unrecoverable even though the element was seen. */
  truncated?: boolean;
  /** Machine-readable reason accompanying `aged-out` or `truncated`. */
  cause?: string;
}

/** True when the element has bytes a resolver could serve. */
export function mediaHasBytes(ref: MediaRef): boolean {
  return ref.source === 'captured' && !ref.truncated && Boolean(ref.locator);
}

/**
 * Media types that must never be rendered inline regardless of what the
 * wire declared or a sniffer concluded. SVG and HTML carry script; XML
 * and PDF have their own active-content histories. A captured body is
 * attacker-influenced content, so it is offered for download and never
 * given a rendering context.
 */
export const NEVER_INLINE_MIMES: readonly string[] = [
  'image/svg+xml',
  'text/html',
  'application/xhtml+xml',
  'text/xml',
  'application/xml',
  'application/pdf',
];

/** True when `mime` must be downloaded rather than previewed. */
export function isNeverInlineMime(mime: string | undefined): boolean {
  if (!mime) return false;
  const bare = mime.split(';')[0].trim().toLowerCase();
  return NEVER_INLINE_MIMES.includes(bare);
}

/** True when the card should attempt an inline preview. */
export function mediaIsPreviewable(ref: MediaRef): boolean {
  if (!mediaHasBytes(ref)) return false;
  if (isNeverInlineMime(ref.mime)) return false;
  return ref.modality === 'image' || ref.modality === 'audio' || ref.modality === 'video';
}

/**
 * Custody state → i18n key suffix. Both UIs render the same six states and
 * both need words rather than slugs, so both were carrying their own copy of
 * this table — and the copies had already diverged: the Agent Dashboard's
 * still listed `redacted`, a state deleted for having no producer, while the
 * Control Plane's did not. One concept, two shapes, drifting.
 *
 * Exhaustive over MediaSource by construction: the Record key type makes a
 * new custody state a compile error here rather than a missing string at
 * runtime on whichever surface was forgotten.
 */
export const MEDIA_STATE_I18N_KEY: Record<MediaSource, string> = {
  captured: 'captured',
  external: 'external',
  'provider-ref': 'providerRef',
  fingerprint: 'fingerprint',
  'aged-out': 'agedOut',
  absent: 'absent',
};

File: packages/shared/transport/normalize/core/media.go
// Media is one concept with one shape: a reference that names WHAT the bytes
// are, PROVES which bytes they were, and — when they are recoverable — says
// where to find them. It never carries content in any encoding.
//
// Split out of types.go because it had grown into a subject of its own, and
// the file-size ratchet is the repo's way of noticing that before a file
// becomes the place everything goes.

package core

// MediaRef describes one media element of a captured exchange. It never
// carries content bytes in any encoding: the bytes already live exactly
// once inside the captured, redaction-governed body, and Locator makes
// them addressable there. Exactly one of Locator / URL / ProviderRef is
// set, matching Source; every other Source sets none of them.
type MediaRef struct {
	// Modality is "image", "audio", "video" or "file".
	Modality string `json:"modality"`
	// Mime is the wire-declared type (data-URI prefix, media_type,
	// mimeType). It is a display hint ONLY — never use it to construct a
	// Blob type, because the wire declares it and a caller controls the
	// wire. Empty means unknown.
	Mime string `json:"mime,omitempty"`
	// SizeBytes is the decoded size, computed arithmetically from the
	// base64 length. Normalizing never decodes to learn it.
	//
	// Deliberately NOT omitempty: a genuinely empty element and an element
	// whose size is unknown are different facts, and collapsing them would
	// make a zero-byte payload indistinguishable from a missing one — the
	// exact ambiguity that let a size of 0 read as "fine" during the
	// investigation.
	SizeBytes int64 `json:"sizeBytes"`
	// Source is the custody state: captured, external, provider-ref,
	// fingerprint, aged-out, redacted or absent. Only "captured" has
	// reachable bytes, and it is the only one a download is offered for.
	Source string `json:"source"`
	// Locator addresses the bytes inside the captured body of the same
	// direction. Set only when Source is "captured".
	Locator string `json:"locator,omitempty"`
	// URL is an inert remote reference the system never fetches.
	URL string `json:"url,omitempty"`
	// ProviderRef is a provider-held file id or file URI.
	ProviderRef string `json:"providerRef,omitempty"`
	// SHA256 is a real hex digest or empty — never a payload prefix.
	SHA256 string `json:"sha256,omitempty"`
	// Truncated marks bytes as unrecoverable: capture cut before this
	// element fully landed, or the captured element fails the decode
	// validator. Either way no download is offered.
	Truncated bool `json:"truncated,omitempty"`
	// Cause names, machine-readably, why bytes are unreachable despite
	// evidence they existed — the classified spill failure for aged-out,
	// the validator failure class for truncated.
	Cause string `json:"cause,omitempty"`
}

// Media custody states.
//
// Six, and every one has a producer. A seventh, "redacted", was defined
// alongside these and nothing ever emitted it: compliance redacts text, not
// media. A custody value nothing can reach is a distinction the system
// claims to make and does not, so it is gone until media redaction exists
// and can set it.
//
// MediaCaptured is the only state with reachable bytes, and even it yields
// them only while Truncated is false — the pair is what mediaHasBytes reads,
// and what decides whether a card offers a control at all.
const (
	MediaCaptured    = "captured"
	MediaExternal    = "external"
	MediaProviderRef = "provider-ref"
	MediaFingerprint = "fingerprint"
	MediaAgedOut     = "aged-out"
	MediaAbsent      = "absent"
)

// Media modalities.
const (
	ModalityImage = "image"
	ModalityAudio = "audio"
	ModalityVideo = "video"
	ModalityFile  = "file"
)
